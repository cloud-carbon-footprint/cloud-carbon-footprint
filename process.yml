# Orb 'circleci/node@4.0.0' resolved to 'circleci/node@4.0.0'
# Orb 'circleci/gcp-gcr@0.8.0' resolved to 'circleci/gcp-gcr@0.8.0'
# Orb 'circleci/gcp-cli@1.8.4' resolved to 'circleci/gcp-cli@1.8.4'
version: 2
jobs:
  deploy:
    docker:
    - image: hashicorp/terraform:0.12.28
    steps:
    - checkout
    - run:
        name: install gcloud
        command: |
          apk add --update python3 py-crcmod bash libc6-compat curl
          export PATH=$PATH:/root/google-cloud-sdk/bin
          curl -sSL https://sdk.cloud.google.com > /tmp/gcloud-install && bash /tmp/gcloud-install --disable-prompts --install-dir=/root
    - run:
        command: |
          IFS="," read -ra PARAMS <<< "GCLOUD_SERVICE_KEY,GOOGLE_PROJECT_ID,GOOGLE_COMPUTE_ZONE"

          for i in "${PARAMS[@]}"; do
            if [[ -z "${i}" ]]; then
              echo "ERROR: Missing environment variable {i}" >&2

              if [[ -n "" ]]; then
                echo "" >&2
              fi

              exit 1
            else
              echo "Yes, ${i} is defined!"
            fi
          done
        name: Checking whether required env vars are set to initialize gcloud CLI...
    - run:
        command: |
          # Store service account
          echo $GCLOUD_SERVICE_KEY > ${HOME}/gcloud-service-key.json

          # Initialize gcloud CLI
          gcloud auth activate-service-account --key-file=${HOME}/gcloud-service-key.json
          gcloud --quiet config set project $GOOGLE_PROJECT_ID
          gcloud --quiet config set compute/zone $GOOGLE_COMPUTE_ZONE
        name: Initialize gcloud CLI to connect to Google Cloud
workflows:
  version: 2
  build-and-deploy:
    jobs:
    - deploy

# Original config.yml file:
# version: 2.1
# orbs:
#   node: circleci/node@4.0.0
#   gcr: circleci/gcp-gcr@0.8.0
#   gcp-cli: circleci/gcp-cli@1.8.4
# jobs:
#   build:
#     working_directory: ~/cloud-carbon-footprint
#     docker:
#       - image: circleci/node:current
#     environment:
#       MAJOR_REVISION: 0
#       MINOR_REVISION: 0
#       GIT_HASH: <<pipeline.git.revision>>
#     # Currently, the cache is commented out since it's slower than installing dependencies every time. Once the
#     # dependency list grows and build time increases, we can start using the cache again. Caching takes about ~30s to
#     # resolve and use as of this comment.
#     steps:
#       - checkout
#       # - restore_cache:
#       #     key: dependency-cache-{{ checksum \"package-lock.json\" }}-{{ checksum \"client/package-lock.json\" }}-{{ checksum \"server/package-lock.json\" }}
#       - run:
#           name: setup
#           command: \"[ ! -d node_modules ] && npm run bootstrap || echo NPM packages unchanged. Using cache.\"
#       # - save_cache:
#       #     key: dependency-cache-{{ checksum \"package-lock.json\" }}-{{ checksum \"client/package-lock.json\" }}-{{ checksum \"server/package-lock.json\" }}
#       #     paths:
#       #       - ./node_modules
#       #       - ./client/node_modules
#       #       - ./server/node_modules
#       - run:
#           name: lint
#           command: 'npm run lint'
#       - run:
#           name: test
#           command: 'npm test'
#   publish-client:
#     description: Build and push client image to Google Container Registry
#     machine:
#       image: ubuntu-1604:202007-01
#     steps:
#       - checkout
#       - run:
#           name: build
#           command: |
#             cd client
#             npm install
#             npm run-script build
#       - gcr/gcr-auth
#       - gcr/build-image:
#           docker-context: client
#           path:  client
#           image: client
#           tag: 0.0.<<pipeline.git.revision>>
#       - gcr/push-image:
#           image: client
#           tag: 0.0.<<pipeline.git.revision>>
#   publish-server:
#     description: Build and push server image to Google Container Registry
#     machine:
#       image: ubuntu-1604:202007-01
#     steps:
#       - checkout
#       - run:
#           name: build
#           command: |
#             cd server
#             npm install
#             npm run-script build
#       - gcr/gcr-auth
#       - gcr/build-image:
#           docker-context: server
#           path:  server
#           image: server
#           tag: 0.0.<<pipeline.git.revision>>
#       - gcr/push-image:
#           image: server
#           tag: 0.0.<<pipeline.git.revision>>
#   deploy:
#     description: Deploy app to gcp
#     docker:
#       - image: hashicorp/terraform:0.12.28
#     steps:
#       - checkout
#       - run: 
#           name: install gcloud
#           command: |
#             apk add --update python3 py-crcmod bash libc6-compat curl
#             export PATH=$PATH:/root/google-cloud-sdk/bin
#             curl -sSL https://sdk.cloud.google.com > /tmp/gcloud-install && bash /tmp/gcloud-install --disable-prompts --install-dir=/root
#       - gcp-cli/initialize
#       # - run:
#       #     environment:
#       #       TF_VAR_server_version: 0.0.<<pipeline.git.revision>>
#       #       TF_VAR_client_version: 0.0.<<pipeline.git.revision>>
#       #     name: deploy
#       #     command: |
# 
#       #       cd terraform
#       #       ./run.sh apply
# workflows:
#   version: 2
#   build-and-deploy:
#     jobs:
#       # - build
#       # - publish-client:
#       #     requires:
#       #       - build
#       # - publish-server:
#       #     requires:
#       #       - build
#       - deploy
#           # requires:
#           #   - publish-client
#           #   - publish-server