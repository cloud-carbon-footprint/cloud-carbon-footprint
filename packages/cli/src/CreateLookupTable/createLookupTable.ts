/*
 * Â© 2021 Thoughtworks, Inc.
 */
import process from 'process'
import commander from 'commander'
import csv from 'csvtojson'
import { createObjectCsvWriter } from 'csv-writer'
import path from 'path'

import { App } from '@cloud-carbon-footprint/app'
import {
  LookupTableInput,
  LookupTableOutput,
} from '@cloud-carbon-footprint/common'
import { validateInputData } from './validateInputData'

export default async function createLookupTable(
  argv: string[] = process.argv,
): Promise<void> {
  const program = new commander.Command()
  program.storeOptionsAsProperties(false)

  program
    .option('--awsInput <filename>', 'File containing relevant AWS input data')
    .option(
      '--awsOutput <filename>',
      'File to write the relevant AWS output data',
      'aws_lookup_data.csv',
    )

  program.parse(argv)

  const programOptions = program.opts()
  const awsInputFile = programOptions.awsInput
  const awsOutputFile = path.join(process.cwd(), programOptions.awsOutput)
  const inputData: LookupTableInput[] = await csv().fromFile(awsInputFile)

  validateInputData(inputData)

  const estimatesData: LookupTableOutput[] =
    await new App().getEstimatesFromInputData(inputData)

  const csvWriter = createObjectCsvWriter({
    path: awsOutputFile,
    header: [
      { id: 'serviceName', title: 'serviceName' },
      { id: 'region', title: 'region' },
      { id: 'usageType', title: 'usageType' },
      { id: 'usageUnit', title: 'usageUnit' },
      { id: 'vCpus', title: 'vCpus' },
      { id: 'kilowattHours', title: 'kilowattHours' },
      { id: 'co2e', title: 'co2e' },
    ],
  })
  await csvWriter.writeRecords(estimatesData)
}
