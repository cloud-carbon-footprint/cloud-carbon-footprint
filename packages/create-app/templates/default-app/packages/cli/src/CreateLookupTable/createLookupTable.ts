/*
 * Â© 2021 Thoughtworks, Inc.
 */
import process from 'process'
import commander from 'commander'
import csv from 'csvtojson'
import path from 'path'

import { App } from '@cloud-carbon-footprint/app'
import {
  LookupTableInput,
  LookupTableOutput,
} from '@cloud-carbon-footprint/common'
import { validateInputData } from './validateInputData'
import { writeToCsv } from './writeToCsv'

export default async function createLookupTable(
  argv: string[] = process.argv,
): Promise<void> {
  const program = new commander.Command()
  program.storeOptionsAsProperties(false)

  program
    .option('--awsInput <filename>', 'File containing relevant AWS input data')
    .option('--gcpInput <filename>', 'File containing relevant GCP input data')
    .option(
      '--azureInput <filename>',
      'File containing relevant Azure input data',
    )
    .option(
      '--awsOutput <filename>',
      'File to write the relevant AWS output data',
      'aws_lookup_data.csv',
    )
    .option(
      '--gcpOutput <filename>',
      'File to write the relevant GCP output data',
      'gcp_lookup_data.csv',
    )
    .option(
      '--azureOutput <filename>',
      'File to write the relevant Azure output data',
      'azure_lookup_data.csv',
    )

  program.parse(argv)

  const programOptions = program.opts()

  const awsInputFile = programOptions.awsInput
  const awsOutputFile = path.join(process.cwd(), programOptions.awsOutput)

  const gcpInputFile = programOptions.gcpInput
  const gcpOutputFile = path.join(process.cwd(), programOptions.gcpOutput)

  const azureInputFile = programOptions.azureInput
  const azureOutputFile = path.join(process.cwd(), programOptions.azureOutput)

  if (awsInputFile) {
    const awsInputData: LookupTableInput[] = await csv().fromFile(awsInputFile)
    validateInputData(awsInputData)
    const awsEstimatesData: LookupTableOutput[] =
      await new App().getAwsEstimatesFromInputData(awsInputData)
    await writeToCsv(awsOutputFile, awsEstimatesData)
  }

  if (gcpInputFile) {
    const gcpInputData: LookupTableInput[] = await csv().fromFile(gcpInputFile)
    validateInputData(gcpInputData)
    const gcpEstimatesData: LookupTableOutput[] =
      await new App().getGcpEstimatesFromInputData(gcpInputData)
    await writeToCsv(gcpOutputFile, gcpEstimatesData)
  }

  if (azureInputFile) {
    const azureInputData: LookupTableInput[] = await csv().fromFile(
      azureInputFile,
    )
    validateInputData(azureInputData)
    const azureEstimatesData: LookupTableOutput[] =
      await new App().getAzureEstimatesFromInputData(azureInputData)
    await writeToCsv(azureOutputFile, azureEstimatesData)
  }
}
