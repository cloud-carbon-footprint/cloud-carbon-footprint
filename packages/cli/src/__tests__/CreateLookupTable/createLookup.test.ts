/*
 * Â© 2021 Thoughtworks, Inc.
 */

import fs from 'fs'
import path from 'path'

import createLookupTable from '../../CreateLookupTable/createLookupTable'
import process from 'process'

describe('createLookupTable', () => {
  let inputFilePath: string
  let outputFilePath: string

  describe('creates lookup table CSV', () => {
    afterEach(() => {
      try {
        fs.unlinkSync(outputFilePath)
      } catch (err) {
        console.error(err)
      }
    })
    it('creates AWS lookup table CSV file, with default output file path', async () => {
      inputFilePath = path.join(__dirname, 'aws_input.test.csv')
      outputFilePath = path.join(process.cwd(), 'aws_lookup_data.csv')

      await createLookupTable([
        'executable',
        'file',
        '--awsInput',
        inputFilePath,
      ])

      expect(fs.existsSync(outputFilePath)).toBe(true)
      expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot()
    })

    it('creates AWS lookup table CSV file, with provided output file name', async () => {
      inputFilePath = path.join(__dirname, 'aws_input.test.csv')
      outputFilePath = path.join(process.cwd(), 'aws_lookup_data.csv')

      await createLookupTable([
        'executable',
        'file',
        '--awsInput',
        inputFilePath,
        '--awsOutput',
        'aws_lookup_data.csv',
      ])

      expect(fs.existsSync(outputFilePath)).toBe(true)
      expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot()
    })
  })

  it('Throws if the input data file is incorrect', async () => {
    inputFilePath = path.join(__dirname, 'incorrect_input.test.csv')

    await expect(() =>
      createLookupTable(['executable', 'file', '--awsInput', inputFilePath]),
    ).rejects.toThrow(
      'Input data is incorrect. Please check your input data file and try again.',
    )
  })
})
