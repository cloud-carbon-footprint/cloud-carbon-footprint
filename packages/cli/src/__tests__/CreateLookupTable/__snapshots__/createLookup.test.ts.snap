// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLookupTable creates lookup table CSV creates AWS lookup table CSV file, with default output file path 1`] = `
"serviceName,region,usageType,usageUnit,vCpus,kilowattHours,co2e
AmazonEC2,us-east-1,USE1-EBSOptimized:m4.xlarge,Hrs,4,0.0172739509,0.0000071817314564295
AmazonEC2,ap-south-1,APS3-EUW3-AWS-Out-Bytes,GB,,0.001135,8.035799999999999e-7
AmazonS3,ap-southeast-2,APS2-TimedStorage-ByteHrs,GB-Mo,,2.0129391486989334e-15,1.5902219274721573e-18
"
`;

exports[`createLookupTable creates lookup table CSV creates AWS lookup table CSV file, with provided output file name 1`] = `
"serviceName,region,usageType,usageUnit,vCpus,kilowattHours,co2e
AmazonEC2,us-east-1,USE1-EBSOptimized:m4.xlarge,Hrs,4,0.0172739509,0.0000071817314564295
AmazonEC2,ap-south-1,APS3-EUW3-AWS-Out-Bytes,GB,,0.001135,8.035799999999999e-7
AmazonS3,ap-southeast-2,APS2-TimedStorage-ByteHrs,GB-Mo,,2.0129391486989334e-15,1.5902219274721573e-18
"
`;
